File: sdk/chocobase-ex/lib/choco_base/postgrest.ex
defmodule ChocoBase.Postgrest do
  @moduledoc """
  PostgREST query builder for Elixir.
  """

  defstruct [:client, :table, :params]

  def new(%ChocoBase{} = client, table) do
    %__MODULE__{
      client: client,
      table: table,
      params: %{}
    }
  end

  def select(%__MODULE__{} = q, columns \\ "*") do
    %{q | params: Map.put(q.params, "select", columns)}
  end

  def eq(%__MODULE__{} = q, column, value) do
    %{q | params: Map.put(q.params, to_string(column), "eq." <> to_string(value))}
  end

  def limit(%__MODULE__{} = q, count) do
    %{q | params: Map.put(q.params, "limit", to_string(count))}
  end

  def execute(%__MODULE__{client: client, table: table, params: params}) do
    query = URI.encode_query(params)
    url = client.url <> "/rest/v1/" <> table <> if(query != "", do: "?" <> query, else: "")
    char_url = String.to_charlist(url)
    char_headers = Enum.map(client.headers, fn {k, v} -> {String.to_charlist(k), String.to_charlist(v)} end)

    case :httpc.request(:get, {char_url, char_headers}, [], []) do
      {:ok, {{_, 200, _}, _resp_headers, resp_body}} ->
        decoded = Jason.decode!(to_string(resp_body))
        {:ok, if(is_map(decoded) and Map.has_key?(decoded, "rows"), do: decoded["rows"], else: decoded)}

      {:ok, {{_, status, _}, _resp_headers, resp_body}} ->
        {:error, {status, to_string(resp_body)}}

      {:error, reason} ->
        {:error, reason}
    end
  end
end

File: sdk/chocobase-php/src/Postgrest.php
<?php

namespace ChocoBase;

class Postgrest
{
    private string $baseUrl;
    private string $table;
    private array $headers;
    private array $params = [];

    public function __construct(string $baseUrl, string $table, array $headers)
    {
        $this->baseUrl = $baseUrl;
        $this->table = $table;
        $this->headers = $headers;
    }

    public function select(string $columns = '*'): self
    {
        $this->params['select'] = $columns;
        return $this;
    }

    public function eq(string $column, $value): self
    {
        $this->params[$column] = "eq.{$value}";
        return $this;
    }

    public function limit(int $count): self
    {
        $this->params['limit'] = (string)$count;
        return $this;
    }

    public function execute(): array
    {
        $query = http_build_query($this->params);
        $url = "{$this->baseUrl}/rest/v1/{$this->table}" . ($query ? "?{$query}" : "");

        $ch = curl_init($url);
        $reqHeaders = [];
        foreach ($this->headers as $k => $v) {
            $reqHeaders[] = "{$k}: {$v}";
        }

        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $reqHeaders
        ]);

        $res = curl_exec($ch);
        curl_close($ch);

        $decoded = json_decode($res ?: '[]', true);
        if (isset($decoded['rows'])) {
            return $decoded['rows'];
        }
        return is_array($decoded) ? $decoded : [];
    }
}
